File: lib/py/euler_tour.py
import sys
sys.setrecursionlimit(10**6)

class EulerTour:
    def __init__(self, n, adj, root=1):
        self.in_time = [0] * (n + 1)
        self.out_time = [0] * (n + 1)
        self.timer = 0
        self.adj = adj
        self._dfs(root, 0)
        
    def _dfs(self, cur, p):
        self.timer += 1
        self.in_time[cur] = self.timer
        for nxt in self.adj[cur]:
            if nxt != p:
                self._dfs(nxt, cur)
        self.out_time[cur] = self.timer
        
    def is_ancestor(self, u, v):
        return self.in_time[u] <= self.in_time[v] and \
               self.out_time[v] <= self.out_time[u]

File: lib/cpp/euler_tour.cpp
#include <iostream>
#include <vector>


using namespace std;

const int MAX = 100005;
vector<int> adj[MAX];
int in_time[MAX], out_time[MAX];
int timer;

void dfs_euler(int cur, int p) {
  in_time[cur] = ++timer;
  for (int next : adj[cur]) {
    if (next != p) {
      dfs_euler(next, cur);
    }
  }
  out_time[cur] = timer;
}

// 쿼리: u가 v의 조상인가? (in[u] <= in[v] && out[v] <= out[u])
bool is_ancestor(int u, int v) {
  return in_time[u] <= in_time[v] && out_time[v] <= out_time[u];
}

// 서브트리 쿼리: [in[u], out[u]] 구간 쿼리로 변환
